lib/std/zig/system/x86.zig

```diff
@@ -0,0 +1,664 @@
+            return;
+        },
+        15 => {
+            if(hasFeature(cpu, .sse3)) {
+                cpu.model = &Target.x86.cpu.k8_sse3;
+                return;
+            }
+            cpu.model = &Target.x86.cpu.k8;
+            return;
+        },
+        16 => {
+            cpu.model = &Target.x86.cpu.amdfam10;
+            return;
+        },
+        20 => {
+            cpu.model = &Target.x86.cpu.btver1;
+            return;
+        },
+        21 => {
+            cpu.model = &Target.x86.cpu.bdver1;
+            if(model >= 0x60 and model <= 0x7f) {
+                cpu.model = &Target.x86.cpu.bdver4;
+                return;
+            }
+            if(model >= 0x30 and model <= 0x3f) {
+                cpu.model = &Target.x86.cpu.bdver3;
+                return;
+            }
+            if((model >= 0x10 and model <= 0x1f) or model == 0x02) {
+                cpu.model = &Target.x86.cpu.bdver2;
+                return;
+            }
+            return;
+        },
+        22 => {
+            cpu.model = &Target.x86.cpu.btver2;
+            return;
+        },
+        23 => {
+            cpu.model = &Target.x86.cpu.znver1;
+            if((model >= 0x30 and model <= 0x3f) or model == 0x71) {
+                cpu.model = &Target.x86.cpu.znver2;
+                return;
+            }
+            return;
+        },
+        else => {
+            cpu.model = &Target.x86.cpu.generic;
+            return;
+        }
+    }
+}
+
+fn detectNativeFeatures(cpu: *Target.Cpu) void {
+
+    var leaf = cpuid(0, 0);
+
+    const max_level = leaf.eax;
+    
+    leaf = cpuid(1, 0);
+
+    setFeature(cpu, .cx8,    bit(leaf.edx,  8));
+    setFeature(cpu, .cx8,    bit(leaf.edx,  8));
+    setFeature(cpu, .cmov,   bit(leaf.edx, 15));
+    setFeature(cpu, .mmx,    bit(leaf.edx, 23));
+    setFeature(cpu, .fxsr,   bit(leaf.edx, 24));
+    setFeature(cpu, .sse,    bit(leaf.edx, 25));
+    setFeature(cpu, .sse2,   bit(leaf.edx, 26));
+    setFeature(cpu, .sse3,   bit(leaf.ecx,  0));
+    setFeature(cpu, .pclmul, bit(leaf.ecx,  1));
+    setFeature(cpu, .ssse3,  bit(leaf.ecx,  9));
+    setFeature(cpu, .cx16,   bit(leaf.ecx, 13));
+    setFeature(cpu, .sse4_1, bit(leaf.ecx, 19));
+    setFeature(cpu, .sse4_2, bit(leaf.ecx, 20));
+    setFeature(cpu, .movbe,  bit(leaf.ecx, 22));
+    setFeature(cpu, .popcnt, bit(leaf.ecx, 23));
+    setFeature(cpu, .aes,    bit(leaf.ecx, 25));
+    setFeature(cpu, .rdrnd,  bit(leaf.ecx, 30));
+
+    const has_avx_save = bit(leaf.ecx, 27) and
+                         bit(leaf.ecx, 28) and
+                         ((leaf.eax & 0x6) == 0x6);
+
+    // Darwin lazily saves the AVX512 context on first use: trust that the OS will
+    // save the AVX512 context if we use AVX512 instructions, even the bit is not
+    // set right now.
+    const has_avx512_save = switch(Target.current.isDarwin()) {
+        true  => true,
+        false => has_avx_save and ((leaf.eax & 0xE0) == 0xE0),
+    };
```

Comment:
I think we cannot rely on this for native CPU detection because users are allowed to do this for example:

`-target native-windows-gnu`

On Darwin, this could provide a false positive for  has_avx512_save, when the cross compiled exe is run in wine for example.

I think it will be better to have this code report the CPU feature accurately, and then userland code elsewhere (not in this pull request) can do the Darwin check if they care about this CPU feature.

However, it looks like `has_avx512_save` is used many times below, and so has an affect on whether other features are available. Another direction to go with this would be to have native CPU detection take an OS parameter as input. Then the switch could be on that parameter rather than the native OS (which may not match the target OS).

Reply:
`has_avx512_save` is used to determine feature availability, which seems like it should be a CPU-level thing, so I'm not sure *how* the operating system would play a role here, but apparently it does.

I've changed this to use the target operating system, instead of the current one, although this will probably need to be tested to ensure it's not passing incorrect values on Darwin systems.